google/cloud/forseti/scanner/audit/groups_settings_rules_engine.py

```diff
@@ -357,21 +354,18 @@ def find_blacklist_violation(self, settings):
             settings (GroupsSettings): Groups Settings.
 
         Returns:
-            str: Statement of what the broken rule required,
-                or empty string in case that rule is not violated.
+             list: List of group settings in violation.
         """
-        has_violation = False
+        violation_settings = []
         if not self.rule['settings']:
-            return has_violation
+            return violation_settings
 
-        violates_every_setting_in_rule = True
         for setting, value in self.rule['settings'].items():
             if getattr(settings, setting) != value:
-                violates_every_setting_in_rule = False
-        if violates_every_setting_in_rule:
-            has_violation = True
+                return []
```

Comment:
That is the current logic that was part of this rules engine, no change in logic. Before it would run through all settings and if any of the settings did not match, then it would not throw a violation. That is how these rules are setup, it will only throw if all the settings in the blacklist rule match. [Definition in the rules file](https://github.com/forseti-security/forseti-security/blob/dev/rules/groups_settings_rules.yaml#L20).

Reply:
I was thinking the old one has a bug but seems like intended behavior based on the documentation